File: cpp/basic/edge_detection.cpp
#include <basic/edge_detection.hpp>
#include <immintrin.h>  // 用于SIMD指令(AVX2)
#include <omp.h>        // 用于OpenMP并行计算
#include <cmath>

namespace ip101 {

using namespace cv;  // 在命名空间开始处声明使用cv命名空间

namespace {
// 内部常量定义
constexpr int CACHE_LINE = 64;    // CPU缓存行大小(字节)
constexpr int SIMD_WIDTH = 32;    // AVX2 SIMD向量宽度(字节)
constexpr int BLOCK_SIZE = 16;    // 分块处理大小，用于优化缓存访问

/**
 * @brief 内存对齐辅助函数
 * @param ptr 需要对齐的指针
 * @param align 对齐字节数，默认为缓存行大小
 * @return 对齐后的指针
 * @note 用于确保内存访问对齐到缓存行边界，提高内存访问效率
 */
inline uchar* alignPtr(uchar* ptr, size_t align = CACHE_LINE) {
    return (uchar*)(((size_t)ptr + align - 1) & ~(align - 1));
}

/**
 * @brief SIMD优化的边缘检测核心计算函数
 * @param padded 填充后的输入图像
 * @param result 输出结果图像
 * @param kernel_x x方向卷积核
 * @param kernel_y y方向卷积核
 * @param y 当前处理的y坐标
 * @param x 当前处理的x坐标
 * @param kernel_size 卷积核大小
 * @note 使用AVX2指令集优化3x3卷积运算，对于其他大小使用普通实现
 */
void process_block_simd(const Mat& padded, Mat& result,
                       const Mat& kernel_x, const Mat& kernel_y,
                       int y, int x, int kernel_size) {
    float gx = 0.0f, gy = 0.0f;

    // 使用AVX2进行并行计算
    if (kernel_size == 3) {
        // 初始化向量寄存器
        __m256 sum_x = _mm256_setzero_ps();  // 8个单精度浮点数的向量
        __m256 sum_y = _mm256_setzero_ps();

        // 3x3卷积核计算
        for (int ky = 0; ky < 3; ++ky) {
            for (int kx = 0; kx < 3; ++kx) {
                float val = padded.at<uchar>(y + ky, x + kx);
                float kx_val = kernel_x.at<float>(ky, kx);
                float ky_val = kernel_y.at<float>(ky, kx);

                // 向量乘法和累加
                sum_x = _mm256_add_ps(sum_x, _mm256_mul_ps(_mm256_set1_ps(val), _mm256_set1_ps(kx_val)));
                sum_y = _mm256_add_ps(sum_y, _mm256_mul_ps(_mm256_set1_ps(val), _mm256_set1_ps(ky_val)));
            }
        }

        // 水平相加得到最终结果
        float gx_arr[8], gy_arr[8];
        _mm256_store_ps(gx_arr, sum_x);
        _mm256_store_ps(gy_arr, sum_y);

        for (int i = 0; i < 8; ++i) {
            gx += gx_arr[i];
            gy += gy_arr[i];
        }
    } else {
        // 对于非3x3的kernel使用普通实现
        for (int ky = 0; ky < kernel_size; ++ky) {
            for (int kx = 0; kx < kernel_size; ++kx) {
                float val = padded.at<uchar>(y + ky, x + kx);
                gx += val * kernel_x.at<float>(ky, kx);
                gy += val * kernel_y.at<float>(ky, kx);
            }
        }
    }

    // 计算梯度幅值并饱和到uchar范围
    result.at<uchar>(y, x) = saturate_cast<uchar>(std::sqrt(gx * gx + gy * gy));
}

// 内存对齐辅助函数
inline size_t get_aligned_size(size_t size, size_t alignment) {
    return (size + alignment - 1) & ~(alignment - 1);
}

// SIMD优化的像素处理
inline void process_pixels_avx2(__m256i* dst, const __m256i* src, const float* kernel, int kernel_size) {
    __m256 sum = _mm256_setzero_ps();
    for (int i = 0; i < kernel_size; ++i) {
        __m256 val = _mm256_cvtepi32_ps(_mm256_loadu_si256(src + i));
        __m256 k = _mm256_broadcast_ss(&kernel[i]);
        sum = _mm256_fmadd_ps(val, k, sum);
    }
    *dst = _mm256_cvtps_epi32(sum);
}

} // anonymous namespace

/**
 * @brief 微分滤波边缘检测实现
 */
void differential_filter(const cv::Mat& src, cv::Mat& dst, int dx, int dy, int ksize) {
    CV_Assert(!src.empty() && src.type() == CV_8UC1);

    dst = Mat::zeros(src.size(), CV_8UC1);
    int pad = ksize / 2;

    // 边缘填充，使用边缘像素值填充
    Mat padded;
    copyMakeBorder(src, padded, pad, pad, pad, pad, BORDER_REPLICATE);

    // 定义微分算子
    Mat kernel_x = (Mat_<float>(3, 3) << 0, 0, 0, -1, 0, 1, 0, 0, 0);
    Mat kernel_y = (Mat_<float>(3, 3) << 0, -1, 0, 0, 0, 0, 0, 1, 0);

    // 使用OpenMP进行并行计算
    #pragma omp parallel for
    for (int y = 0; y < src.rows; ++y) {
        for (int x = 0; x < src.cols; ++x) {
            process_block_simd(padded, dst, kernel_x, kernel_y, y, x, ksize);
        }
    }
}

/**
 * @brief Sobel算子边缘检测实现
 */
void sobel_filter(const cv::Mat& src, cv::Mat& dst, int dx, int dy, int ksize, double scale) {
    CV_Assert(!src.empty() && src.type() == CV_8UC1);

    dst = Mat::zeros(src.size(), CV_8UC1);
    int pad = ksize / 2;

    // 边缘填充
    Mat padded;
    copyMakeBorder(src, padded, pad, pad, pad, pad, BORDER_REPLICATE);

    // 定义Sobel算子
    Mat kernel_x = (Mat_<float>(3, 3) << -1, 0, 1, -2, 0, 2, -1, 0, 1);
    Mat kernel_y = (Mat_<float>(3, 3) << -1, -2, -1, 0, 0, 0, 1, 2, 1);

    // 使用OpenMP进行并行计算
    #pragma omp parallel for
    for (int y = 0; y < src.rows; ++y) {
        for (int x = 0; x < src.cols; ++x) {
            process_block_simd(padded, dst, kernel_x, kernel_y, y, x, ksize);
        }
    }

    // 应用缩放因子
    if (scale != 1.0) {
        dst = dst * scale;
    }
}

/**
 * @brief Prewitt算子边缘检测实现
 */
void prewitt_filter(const cv::Mat& src, cv::Mat& dst, int dx, int dy) {
    CV_Assert(!src.empty() && src.type() == CV_8UC1);

    dst = Mat::zeros(src.size(), CV_8UC1);
    int ksize = 3; // Prewitt算子固定为3x3
    int pad = ksize / 2;

    // 边缘填充
    Mat padded;
    copyMakeBorder(src, padded, pad, pad, pad, pad, BORDER_REPLICATE);

    // 定义Prewitt算子
    Mat kernel_x = (Mat_<float>(3, 3) << -1, 0, 1, -1, 0, 1, -1, 0, 1);
    Mat kernel_y = (Mat_<float>(3, 3) << -1, -1, -1, 0, 0, 0, 1, 1, 1);

    // 使用OpenMP进行并行计算
    #pragma omp parallel for
    for (int y = 0; y < src.rows; ++y) {
        for (int x = 0; x < src.cols; ++x) {
            process_block_simd(padded, dst, kernel_x, kernel_y, y, x, ksize);
        }
    }
}

/**
 * @brief Laplacian算子边缘检测实现
 */
void laplacian_filter(const cv::Mat& src, cv::Mat& dst, int ksize, double scale) {
    CV_Assert(!src.empty() && src.type() == CV_8UC1);

    dst = Mat::zeros(src.size(), CV_8UC1);
    int pad = ksize / 2;

    // 边缘填充
    Mat padded;
    copyMakeBorder(src, padded, pad, pad, pad, pad, BORDER_REPLICATE);

    // 定义Laplacian算子
    Mat kernel = (Mat_<float>(3, 3) << 0, 1, 0, 1, -4, 1, 0, 1, 0);
    Mat kernel_x = kernel.clone(); // 为了兼容process_block_simd函数
    Mat kernel_y = kernel.clone();

    // 使用OpenMP进行并行计算
    #pragma omp parallel for
    for (int y = 0; y < src.rows; ++y) {
        for (int x = 0; x < src.cols; ++x) {
            float sum = 0.0f;

            // 对于非3x3的kernel使用普通实现
            for (int ky = 0; ky < ksize; ++ky) {
                for (int kx = 0; kx < ksize; ++kx) {
                    float val = padded.at<uchar>(y + ky, x + kx);
                    sum += val * kernel.at<float>(ky % 3, kx % 3); // 使用模运算确保索引在有效范围内
                }
            }

            // 取绝对值并饱和到uchar范围
            dst.at<uchar>(y, x) = saturate_cast<uchar>(std::abs(sum) * scale);
        }
    }
}

/**
 * @brief 浮雕效果实现
 */
void emboss_effect(const cv::Mat& src, cv::Mat& dst, int direction) {
    CV_Assert(!src.empty() && src.type() == CV_8UC1);

    dst = Mat::zeros(src.size(), CV_8UC1);
    int ksize = 3; // 浮雕效果固定使用3x3卷积核
    int pad = ksize / 2;
    int offset = 128; // 默认偏移值

    // 边缘填充
    Mat padded;
    copyMakeBorder(src, padded, pad, pad, pad, pad, BORDER_REPLICATE);

    // 根据方向选择浮雕算子
    Mat kernel;
    switch (direction) {
        case 0: // 默认方向（右下）
            kernel = (Mat_<float>(3, 3) << 2, 0, 0, 0, -1, 0, 0, 0, -1);
            break;
        case 1: // 右
            kernel = (Mat_<float>(3, 3) << 0, 0, 2, 0, -1, 0, 0, 0, -1);
            break;
        case 2: // 右上
            kernel = (Mat_<float>(3, 3) << 0, 0, 2, 0, -1, 0, -1, 0, 0);
            break;
        case 3: // 上
            kernel = (Mat_<float>(3, 3) << 0, 2, 0, 0, -1, 0, 0, -1, 0);
            break;
        case 4: // 左上
            kernel = (Mat_<float>(3, 3) << 2, 0, 0, 0, -1, 0, 0, 0, -1);
            kernel = kernel.t(); // 转置
            break;
        case 5: // 左
            kernel = (Mat_<float>(3, 3) << 0, 0, -1, 0, -1, 0, 2, 0, 0);
            break;
        case 6: // 左下
            kernel = (Mat_<float>(3, 3) << -1, 0, 0, 0, -1, 0, 0, 0, 2);
            break;
        case 7: // 下
            kernel = (Mat_<float>(3, 3) << 0, -1, 0, 0, -1, 0, 0, 2, 0);
            break;
        default:
            kernel = (Mat_<float>(3, 3) << 2, 0, 0, 0, -1, 0, 0, 0, -1);
            break;
    }

    // 使用OpenMP进行并行计算
    #pragma omp parallel for
    for (int y = 0; y < src.rows; ++y) {
        for (int x = 0; x < src.cols; ++x) {
            float sum = 0.0f;

            // 计算卷积
            for (int ky = 0; ky < ksize; ++ky) {
                for (int kx = 0; kx < ksize; ++kx) {
                    float val = padded.at<uchar>(y + ky, x + kx);
                    sum += val * kernel.at<float>(ky, kx);
                }
            }

            // 添加偏移并饱和到uchar范围
            dst.at<uchar>(y, x) = saturate_cast<uchar>(sum + offset);
        }
    }
}

/**
 * @brief 综合边缘检测实现
 */
void edge_detection(const cv::Mat& src, cv::Mat& dst, const std::string& method, double thresh_val) {
    CV_Assert(!src.empty());

    // 转换为灰度图
    Mat gray;
    if (src.channels() == 3) {
        cvtColor(src, gray, COLOR_BGR2GRAY);
    } else {
        gray = src.clone();
    }

    // 根据选择的方法进行边缘检测
    Mat result;
    if (method == "sobel") {
        sobel_filter(gray, result, 1, 1, 3, 1.0); // dx=1, dy=1, ksize=3, scale=1.0
    } else if (method == "prewitt") {
        prewitt_filter(gray, result, 1, 1); // dx=1, dy=1
    } else if (method == "laplacian") {
        laplacian_filter(gray, result, 3, 1.0); // ksize=3, scale=1.0
    } else {
        throw std::invalid_argument("Unsupported method: " + method);
    }

    // 二值化处理
    threshold(result, dst, thresh_val, 255, THRESH_BINARY);
}

} // namespace ip101

File: python/basic/edge_detection.py
"""
边缘检测相关问题：
11. 微分滤波 - 使用3x3微分滤波器进行边缘检测
12. Sobel滤波 - 使用Sobel算子进行边缘检测
13. Prewitt滤波 - 使用Prewitt算子进行边缘检测
14. Laplacian滤波 - 使用Laplacian算子进行边缘检测
15. 浮雕效果 - 使用差分和偏移实现浮雕效果
16. 边缘检测 - 综合多种边缘检测方法
"""

import cv2
import numpy as np

def differential_filter(img_path, kernel_size=3):
    """
    问题11：微分滤波
    使用3x3微分滤波器进行边缘检测

    参数:
        img_path: 输入图像路径
        kernel_size: 滤波器大小，默认为3

    返回:
        边缘检测结果
    """
    # 读取图像
    img = cv2.imread(img_path)
    if img is None:
        raise ValueError(f"无法读取图像: {img_path}")

    # 转换为灰度图
    gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)

    # 获取图像尺寸
    h, w = gray.shape

    # 创建输出图像
    result = np.zeros_like(gray)

    # 计算填充大小
    pad = kernel_size // 2

    # 对图像进行填充
    padded = np.pad(gray, ((pad, pad), (pad, pad)), mode='edge')

    # 手动实现微分滤波
    for y in range(h):
        for x in range(w):
            # 提取当前窗口
            window = padded[y:y+kernel_size, x:x+kernel_size]

            # 计算x方向和y方向的差分
            dx = window[1, 2] - window[1, 0]
            dy = window[2, 1] - window[0, 1]

            # 计算梯度幅值
            result[y, x] = np.sqrt(dx*dx + dy*dy)

    # 归一化到0-255
    result = np.clip(result, 0, 255).astype(np.uint8)

    return result

def sobel_filter(img_path, kernel_size=3):
    """
    问题12：Sobel滤波
    使用Sobel算子进行边缘检测

    参数:
        img_path: 输入图像路径
        kernel_size: 滤波器大小，默认为3

    返回:
        边缘检测结果
    """
    # 读取图像
    img = cv2.imread(img_path)
    if img is None:
        raise ValueError(f"无法读取图像: {img_path}")

    # 转换为灰度图
    gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)

    # 获取图像尺寸
    h, w = gray.shape

    # 创建输出图像
    result = np.zeros_like(gray)

    # 计算填充大小
    pad = kernel_size // 2

    # 对图像进行填充
    padded = np.pad(gray, ((pad, pad), (pad, pad)), mode='edge')

    # 定义Sobel算子
    sobel_x = np.array([[-1, 0, 1], [-2, 0, 2], [-1, 0, 1]])
    sobel_y = np.array([[-1, -2, -1], [0, 0, 0], [1, 2, 1]])

    # 手动实现Sobel滤波
    for y in range(h):
        for x in range(w):
            # 提取当前窗口
            window = padded[y:y+kernel_size, x:x+kernel_size]

            # 计算x方向和y方向的卷积
            gx = np.sum(window * sobel_x)
            gy = np.sum(window * sobel_y)

            # 计算梯度幅值
            result[y, x] = np.sqrt(gx*gx + gy*gy)

    # 归一化到0-255
    result = np.clip(result, 0, 255).astype(np.uint8)

    return result

def prewitt_filter(img_path, kernel_size=3):
    """
    问题13：Prewitt滤波
    使用Prewitt算子进行边缘检测

    参数:
        img_path: 输入图像路径
        kernel_size: 滤波器大小，默认为3

    返回:
        边缘检测结果
    """
    # 读取图像
    img = cv2.imread(img_path)
    if img is None:
        raise ValueError(f"无法读取图像: {img_path}")

    # 转换为灰度图
    gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)

    # 获取图像尺寸
    h, w = gray.shape

    # 创建输出图像
    result = np.zeros_like(gray)

    # 计算填充大小
    pad = kernel_size // 2

    # 对图像进行填充
    padded = np.pad(gray, ((pad, pad), (pad, pad)), mode='edge')

    # 定义Prewitt算子
    prewitt_x = np.array([[-1, 0, 1], [-1, 0, 1], [-1, 0, 1]])
    prewitt_y = np.array([[-1, -1, -1], [0, 0, 0], [1, 1, 1]])

    # 手动实现Prewitt滤波
    for y in range(h):
        for x in range(w):
            # 提取当前窗口
            window = padded[y:y+kernel_size, x:x+kernel_size]

            # 计算x方向和y方向的卷积
            gx = np.sum(window * prewitt_x)
            gy = np.sum(window * prewitt_y)

            # 计算梯度幅值
            result[y, x] = np.sqrt(gx*gx + gy*gy)

    # 归一化到0-255
    result = np.clip(result, 0, 255).astype(np.uint8)

    return result

def laplacian_filter(img_path, kernel_size=3):
    """
    问题14：Laplacian滤波
    使用Laplacian算子进行边缘检测

    参数:
        img_path: 输入图像路径
        kernel_size: 滤波器大小，默认为3

    返回:
        边缘检测结果
    """
    # 读取图像
    img = cv2.imread(img_path)
    if img is None:
        raise ValueError(f"无法读取图像: {img_path}")

    # 转换为灰度图
    gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)

    # 获取图像尺寸
    h, w = gray.shape

    # 创建输出图像
    result = np.zeros_like(gray)

    # 计算填充大小
    pad = kernel_size // 2

    # 对图像进行填充
    padded = np.pad(gray, ((pad, pad), (pad, pad)), mode='edge')

    # 定义Laplacian算子
    laplacian = np.array([[0, 1, 0], [1, -4, 1], [0, 1, 0]])

    # 手动实现Laplacian滤波
    for y in range(h):
        for x in range(w):
            # 提取当前窗口
            window = padded[y:y+kernel_size, x:x+kernel_size]

            # 计算Laplacian卷积
            result[y, x] = np.sum(window * laplacian)

    # 取绝对值并归一化到0-255
    result = np.abs(result)
    result = np.clip(result, 0, 255).astype(np.uint8)

    return result

def emboss_effect(img_path, kernel_size=3, offset=128):
    """
    问题15：浮雕效果
    使用差分和偏移实现浮雕效果

    参数:
        img_path: 输入图像路径
        kernel_size: 滤波器大小，默认为3
        offset: 偏移值，默认为128

    返回:
        浮雕效果图像
    """
    # 读取图像
    img = cv2.imread(img_path)
    if img is None:
        raise ValueError(f"无法读取图像: {img_path}")

    # 转换为灰度图
    gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)

    # 获取图像尺寸
    h, w = gray.shape

    # 创建输出图像
    result = np.zeros_like(gray)

    # 计算填充大小
    pad = kernel_size // 2

    # 对图像进行填充
    padded = np.pad(gray, ((pad, pad), (pad, pad)), mode='edge')

    # 定义浮雕算子
    emboss = np.array([[2, 0, 0], [0, -1, 0], [0, 0, -1]])

    # 手动实现浮雕效果
    for y in range(h):
        for x in range(w):
            # 提取当前窗口
            window = padded[y:y+kernel_size, x:x+kernel_size]

            # 计算浮雕卷积
            result[y, x] = np.sum(window * emboss) + offset

    # 归一化到0-255
    result = np.clip(result, 0, 255).astype(np.uint8)

    return result

def edge_detection(img_path, method='sobel', threshold=100):
    """
    问题16：边缘检测
    综合多种边缘检测方法

    参数:
        img_path: 输入图像路径
        method: 边缘检测方法，可选 'sobel', 'prewitt', 'laplacian'
        threshold: 阈值，默认为100

    返回:
        边缘检测结果
    """
    # 读取图像
    img = cv2.imread(img_path)
    if img is None:
        raise ValueError(f"无法读取图像: {img_path}")

    # 转换为灰度图
    gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)

    # 根据选择的方法进行边缘检测
    if method == 'sobel':
        # 使用Sobel算子
        result = sobel_filter(img_path)
    elif method == 'prewitt':
        # 使用Prewitt算子
        result = prewitt_filter(img_path)
    elif method == 'laplacian':
        # 使用Laplacian算子
        result = laplacian_filter(img_path)
    else:
        raise ValueError(f"不支持的方法: {method}")

    # 二值化处理
    _, binary = cv2.threshold(result, threshold, 255, cv2.THRESH_BINARY)

    return binary

def main(problem_id=11):
    """
    主函数，通过problem_id选择要运行的问题

    参数:
        problem_id: 问题编号，默认为11
    """
    input_path = "../images/imori.jpg"

    if problem_id == 11:
        result = differential_filter(input_path)
        output_path = "../images/answer_11.jpg"
        title = "微分滤波"
    elif problem_id == 12:
        result = sobel_filter(input_path)
        output_path = "../images/answer_12.jpg"
        title = "Sobel滤波"
    elif problem_id == 13:
        result = prewitt_filter(input_path)
        output_path = "../images/answer_13.jpg"
        title = "Prewitt滤波"
    elif problem_id == 14:
        result = laplacian_filter(input_path)
        output_path = "../images/answer_14.jpg"
        title = "Laplacian滤波"
    elif problem_id == 15:
        result = emboss_effect(input_path)
        output_path = "../images/answer_15.jpg"
        title = "浮雕效果"
    elif problem_id == 16:
        result = edge_detection(input_path)
        output_path = "../images/answer_16.jpg"
        title = "边缘检测"
    else:
        print(f"问题 {problem_id} 不存在")
        return

    # 保存结果
    cv2.imwrite(output_path, result)

    # 显示结果
    cv2.imshow(title, result)
    cv2.waitKey(0)
    cv2.destroyAllWindows()

if __name__ == "__main__":
    import sys
    problem_id = int(sys.argv[1]) if len(sys.argv) > 1 else 11
    main(problem_id)